lib/trino-parquet/src/main/java/io/trino/parquet/reader/TrinoColumnIndexStore.java

```diff
@@ -24,167 +25,202 @@
 import org.apache.parquet.internal.column.columnindex.OffsetIndex;
 import org.apache.parquet.internal.filter2.columnindex.ColumnIndexStore;
 import org.apache.parquet.internal.hadoop.metadata.IndexReference;
+import org.apache.parquet.schema.PrimitiveType;
 
 import java.io.IOException;
+import java.util.ArrayList;
 import java.util.HashMap;
+import java.util.List;
 import java.util.Map;
 import java.util.Set;
+import java.util.function.BiConsumer;
 
-import static java.util.Collections.emptySet;
+import static com.google.common.base.Verify.verify;
+import static java.lang.Math.toIntExact;
 
 /**
  * Internal implementation of {@link ColumnIndexStore}.
- * Copy of org.apache.parquet.hadoop.ColumnIndexStoreImpl which is not accessible
+ * Similar to org.apache.parquet.hadoop.ColumnIndexStoreImpl which is not accessible
  */
 public class TrinoColumnIndexStore
         implements ColumnIndexStore
 {
-    private interface IndexStore
-    {
-        ColumnIndex getColumnIndex();
+    private static final Logger log = Logger.get(TrinoColumnIndexStore.class);
 
-        OffsetIndex getOffsetIndex();
+    private final ParquetDataSource dataSource;
+    private final List<ColumnIndexMetadata> columnIndexReferences;
+    private final List<ColumnIndexMetadata> offsetIndexReferences;
+
+    private Map<ColumnPath, ColumnIndex> columnIndexStore;
+    private Map<ColumnPath, OffsetIndex> offsetIndexStore;
+    private boolean columnIndexRead;
+    private boolean offsetIndexRead;
+
+    /**
+     * Creates a column index store which lazily reads column/offset indexes for the columns in paths.
+     *
+     * @param columnsRead is the set of columns used for projection
+     * @param columnsFiltered is the set of columns used for filtering
+     */
+    public static ColumnIndexStore create(
```

Comment:
Is there a reason to use a static creator instead of just exposing the constructor?